Test suite with two failing tests

On the current master (6e89ce655) I have two failing tests in `tests/test_tree.py`, see below.

Should the first one check for a raised `NonImplementedError`, maybe?

```
░ tgal@staticbox:uproot  remotes/upstream/master uproot 
░ 13:52:48 128 > pip freeze           
attrs==19.3.0
awkward==0.12.18
cachetools==4.0.0
certifi==2019.11.28
cffi==1.13.2
chardet==3.0.4
idna==2.8
importlib-metadata==1.3.0
lz4==2.2.1
mock==3.0.5
more-itertools==8.0.2
numpy==1.17.4
packaging==19.2
pandas==0.25.3
pluggy==0.13.1
py==1.8.0
pycparser==2.19
pyparsing==2.4.5
pytest==5.3.2
python-dateutil==2.8.1
pytz==2019.3
requests==2.22.0
six==1.13.0
-e git+git@github.com:tamasgal/uproot.git@6e89ce6556fe204ee4cc9e0c0c06f6d78f97503b#egg=uproot
uproot-methods==0.7.2
urllib3==1.25.7
wcwidth==0.1.7
xxhash==1.4.3
zipp==0.6.0
zstandard==0.12.0
░ tgal@staticbox:uproot  remotes/upstream/master uproot 
░ 13:53:33 > python --version   
Python 3.7.4
```

Test suite run:

```
tests/test_tree.py .F.......F..........................                                                                    [ 87%]
```

Failures:

```python
============================================================ FAILURES ============================================================
______________________________________________ Test.test_branch_auto_interpretation ______________________________________________

self = <tests.test_tree.Test testMethod=test_branch_auto_interpretation>

    def test_branch_auto_interpretation(self):
        # The aim is to reduce this list in a controlled manner
        known_branches_without_interp = [
            b'event',
            b'TObject',
            b'fClosestDistance',
            b'fEvtHdr',
            b'fTracks',
            b'fTracks.fPointValue',
            b'fTriggerBits',
            b'fTriggerBits.TObject'
        ]
        path = os.path.join("tests", "samples", "Event.root")
        if os.path.exists(path):
            tree = uproot.open(path)["T"]
        else:
            tree = uproot.open("http://scikit-hep.org/uproot/examples/Event.root")["T"]
        branches_without_interp = [b.name for b in tree.allvalues() if b.interpretation is None]
        assert branches_without_interp == known_branches_without_interp
        assert tree.array("fTracks.fTArray[3]", entrystop=10)[5][10].tolist()  == [11.03951644897461, 19.40645980834961, 34.54059982299805]
        assert tree.array("fTracks.fCharge", entrystop=10)[0][0:10].tolist()   == [1.0, 1.0, 1.0, 1.0,-1.0, 0.0, 1.0, 0.0, 0.0, 0.0]
>       assert tree.array("fMatrix[4][4]", entrystop=10)[0][1].tolist()        == [-0.13630907237529755, 0.8007842898368835, 1.706235647201538, 0.0]

tests/test_tree.py:634: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
uproot/tree.py:500: in array
    return tbranch.array(interpretation=interpretation, entrystart=entrystart, entrystop=entrystop, flatten=flatten, awkwardlib=awkwardlib, cache=cache, basketcache=basketcache, keycache=keycache, executor=executor, blocking=blocking)
uproot/tree.py:1381: in array
    basket_itemoffset = self._basket_itemoffset(interpretation, basketstart, basketstop, keycache)
uproot/tree.py:1334: in _basket_itemoffset
    for j, key in enumerate(self._threadsafe_iterate_keys(keycache, True, basketstart, basketstop)):
uproot/tree.py:1040: in _threadsafe_iterate_keys
    key = self._basketkey(keysource, i, complete)
uproot/tree.py:1757: in _basketkey
    return self._BasketKey(source.parent(), Cursor(self._fBasketSeek[i]), self.compression, complete)
uproot/tree.py:1635: in __init__
    self._fNbytes, self._fVersion, self._fObjlen, self._fDatime, self._fKeylen, self._fCycle, self._fSeekKey, self._fSeekPdir = cursor.fields(source, TBranchMethods._BasketKey._format_small)
uproot/source/cursor.py:46: in fields
    return format.unpack(source.data(start, stop))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <uproot.source.http.HTTPSource object at 0x7f350fac69d0>, start = 8166160, stop = 8166186, dtype = None

    def data(self, start, stop, dtype=None):
        if dtype is None:
            thedtype = numpy.dtype(numpy.uint8)
        else:
            thedtype = dtype
    
        # assert start >= 0
        # assert stop >= 0
        # assert stop >= start
    
        chunkstart = start // self._chunkbytes
        if stop % self._chunkbytes == 0:
            chunkstop = stop // self._chunkbytes
        else:
            chunkstop = stop // self._chunkbytes + 1
    
        out = numpy.empty((stop - start) // thedtype.itemsize, dtype=thedtype)
    
        for chunkindex in range(chunkstart, chunkstop):
            chunk = None
            if self._futures is not None:
                future = self._futures.pop(chunkindex, None)
                if future is not None:
                    chunk = future.result()
    
            if chunk is None:
                try:
                    chunk = self.cache[chunkindex]
                except KeyError:
                    self._open()
                    chunk = self._read(chunkindex)
    
            cstart = 0
            cstop = self._chunkbytes
            gstart = chunkindex * self._chunkbytes
            gstop = (chunkindex + 1) * self._chunkbytes
    
            if len(chunk) > self._chunkbytes:
                if not numpy.array_equal(chunk[:4], list(b"root")):
>                   raise NotImplementedError("Expected {0} or fewer bytes but received {1} and data does not appear to be an entire ROOT file.".format(self._chunkbytes, len(chunk)))
E                   NotImplementedError: Expected 1048576 or fewer bytes but received 1048577 and data does not appear to be an entire ROOT file.

uproot/source/chunked.py:125: NotImplementedError
_____________________________________________________ Test.test_hist_in_tree _____________________________________________________

self = <tests.test_tree.Test testMethod=test_hist_in_tree>

    def test_hist_in_tree(self):
        path = os.path.join("tests", "samples", "Event.root")
        if os.path.exists(path):
            tree = uproot.open(path)["T"]
        else:
            tree = uproot.open("http://scikit-hep.org/uproot/examples/Event.root")["T"]
        check = [0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0,
                 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 1.0,
                 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
                 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 1.0, 1.0,
                 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 2.0, 0.0, 0.0, 1.0,
                 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.0, 0.0, 1.0, 0.0, 1.0, 0.0, 0.0,
                 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0]
    
>       assert tree.array("fH")[20].values.tolist() == check

tests/test_tree.py:611: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
uproot/tree.py:500: in array
    return tbranch.array(interpretation=interpretation, entrystart=entrystart, entrystop=entrystop, flatten=flatten, awkwardlib=awkwardlib, cache=cache, basketcache=basketcache, keycache=keycache, executor=executor, blocking=blocking)
uproot/tree.py:1381: in array
    basket_itemoffset = self._basket_itemoffset(interpretation, basketstart, basketstop, keycache)
uproot/tree.py:1334: in _basket_itemoffset
    for j, key in enumerate(self._threadsafe_iterate_keys(keycache, True, basketstart, basketstop)):
uproot/tree.py:1040: in _threadsafe_iterate_keys
    key = self._basketkey(keysource, i, complete)
uproot/tree.py:1757: in _basketkey
    return self._BasketKey(source.parent(), Cursor(self._fBasketSeek[i]), self.compression, complete)
uproot/tree.py:1635: in __init__
    self._fNbytes, self._fVersion, self._fObjlen, self._fDatime, self._fKeylen, self._fCycle, self._fSeekKey, self._fSeekPdir = cursor.fields(source, TBranchMethods._BasketKey._format_small)
uproot/source/cursor.py:46: in fields
    return format.unpack(source.data(start, stop))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <uproot.source.http.HTTPSource object at 0x7f3510c6f4d0>, start = 1171681, stop = 1171707, dtype = None

    def data(self, start, stop, dtype=None):
        if dtype is None:
            thedtype = numpy.dtype(numpy.uint8)
        else:
            thedtype = dtype
    
        # assert start >= 0
        # assert stop >= 0
        # assert stop >= start
    
        chunkstart = start // self._chunkbytes
        if stop % self._chunkbytes == 0:
            chunkstop = stop // self._chunkbytes
        else:
            chunkstop = stop // self._chunkbytes + 1
    
        out = numpy.empty((stop - start) // thedtype.itemsize, dtype=thedtype)
    
        for chunkindex in range(chunkstart, chunkstop):
            chunk = None
            if self._futures is not None:
                future = self._futures.pop(chunkindex, None)
                if future is not None:
                    chunk = future.result()
    
            if chunk is None:
                try:
                    chunk = self.cache[chunkindex]
                except KeyError:
                    self._open()
                    chunk = self._read(chunkindex)
    
            cstart = 0
            cstop = self._chunkbytes
            gstart = chunkindex * self._chunkbytes
            gstop = (chunkindex + 1) * self._chunkbytes
    
            if len(chunk) > self._chunkbytes:
                if not numpy.array_equal(chunk[:4], list(b"root")):
                    raise NotImplementedError("Expected {0} or fewer bytes but received {1} and data does not appear to be an entire ROOT file.".format(self._chunkbytes, len(chunk)))
                self.cache = {}
                for i in range(0, len(chunk), self._chunkbytes):
                    self.cache[i // self._chunkbytes] = chunk[i:i+self._chunkbytes]
                chunk = self.cache[chunkindex]
                # Dismiss any pending futures as everything has already been loaded
                self.dismiss()
            else:
                self.cache[chunkindex] = chunk
    
            if gstart < start:
                cstart += start - gstart
                gstart += start - gstart
            if gstop > stop:
                cstop -= gstop - stop
                gstop -= gstop - stop
    
            if cstop - cstart > len(chunk):
>               raise IndexError("indexes {0}:{1} are beyond the end of data source {2}".format(gstart + len(chunk), stop, repr(self.path)))
E               IndexError: indexes 1171682:1171707 are beyond the end of data source 'http://scikit-hep.org/uproot/examples/Event.root'

uproot/source/chunked.py:143: IndexError
=========================================== 2 failed, 105 passed, 2 skipped in 10.92s ============================================
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -44,13 +44,13 @@ install:
   # FIXME: Mamba causes pip install numpy to be extremely slow
   - if [[ "${PYVER}" = pypy* ]]; then
       conda create --quiet --yes -n testenv ${PYVER};
-    elif [ "${PYVER}" = "2.7" ] || [ "${PYVER}" = "3.6" ] || [ "${PYVER}" = "3.7" ]; then
+    elif [ "${PYVER}" = "2.7" ] || [ "${PYVER}" = "3.6" ] || [ "${PYVER}" = "3.7" || [ "${PYVER}" = "3.8" ]; then
       conda create --quiet --yes -n testenv python=${PYVER} pip;
     else
       conda create --quiet --yes -n testenv python=${PYVER};
     fi
   - source activate testenv
-  - if [ "${PYVER}" = "2.7" ] || [ "${PYVER}" = "3.6" ] || [ "${PYVER}" = "3.7" ]; then
+  - if [ "${PYVER}" = "2.7" ] || [ "${PYVER}" = "3.6" ] || [ "${PYVER}" = "3.7" || [ "${PYVER}" = "3.8" ]; then
       conda install --quiet --yes python=${PYVER} pip root;
       source activate testenv;
     fi
@@ -78,7 +78,11 @@ addons:
       - libatlas-base-dev
 
 script:
-  pytest --error-for-skips -v tests
+  if [ "${PYVER}" = "2.7" ] || [ "${PYVER}" = "3.6" ] || [ "${PYVER}" = "3.7" || [ "${PYVER}" = "3.8" ]; then
+    pytest --error-for-skips -v tests;
+  else
+    pytest -v tests;
+  fi
 
 notifications:
   slack: scikit-hep:b6cgBXwccPoaCNLn5VKFJFVy

--- a/.travis.yml
+++ b/.travis.yml
@@ -69,6 +69,7 @@ install:
       conda install -c anaconda python=${PYVER} pyopenssl;
     fi
   - wget -O tests/samples/Event.root http://scikit-hep.org/uproot/examples/Event.root
+  - pip install git+https://github.com/chrisburr/pytest-error-for-skips.git@patch-1
 
 addons:
   apt:
@@ -77,7 +78,7 @@ addons:
       - libatlas-base-dev
 
 script:
-  pytest -v tests
+  pytest --error-for-skips -v tests
 
 notifications:
   slack: scikit-hep:b6cgBXwccPoaCNLn5VKFJFVy

--- a/tests/test_cache.py
+++ b/tests/test_cache.py
@@ -2,13 +2,9 @@
 
 # BSD 3-Clause License; see https://github.com/scikit-hep/uproot/blob/master/LICENSE
 
-import unittest
-
-from collections import namedtuple
-
 import uproot
 
-class Test(unittest.TestCase):
+class Test(object):
     def test_flat_array(self):
         branch = uproot.open("tests/samples/sample-6.10.05-uncompressed.root")["sample"]["i8"]
         expectation = [-15, -14, -13, -12, -11, -10, -9, -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]

--- a/tests/test_compression.py
+++ b/tests/test_compression.py
@@ -2,8 +2,6 @@
 
 # BSD 3-Clause License; see https://github.com/scikit-hep/uproot/blob/master/LICENSE
 
-import unittest
-
 try:
     import lzma
 except ImportError:
@@ -13,7 +11,7 @@ import zstandard
 
 import uproot
 
-class Test(unittest.TestCase):
+class Test(object):
     def test_compression_identity(self):
         assert uproot.open("tests/samples/Zmumu-zlib.root").compression.algoname == "zlib"
         assert uproot.open("tests/samples/Zmumu-zlib.root").compression.level == 4

--- a/tests/test_http.py
+++ b/tests/test_http.py
@@ -2,8 +2,6 @@
 
 # BSD 3-Clause License; see https://github.com/scikit-hep/uproot/blob/master/LICENSE
 
-import unittest
-
 import pytest
 import mock
 from requests.exceptions import HTTPError
@@ -41,7 +39,7 @@ def mock_get_local_instead_of_http(url="", headers={}, auth=None, **kwargs):
         return MockResponse(401)
 
 @mock.patch("requests.get", mock_get_local_instead_of_http)
-class Test(unittest.TestCase):
+class Test(object):
     def test_no_auth_needed_no_auth(self):
         f = uproot.open(URL)
         assert type(f) == uproot.rootio.ROOTDirectory

--- a/tests/test_issues.py
+++ b/tests/test_issues.py
@@ -267,27 +267,19 @@ class Test(object):
         ]
 
     def test_issue232(self):
-        try:
-            import pandas
-        except ImportError:
-            pass
-        else:
-            t = uproot.open("tests/samples/issue232.root")["fTreeV0"]
-            t.pandas.df(
-                ["V0Hyper.fNsigmaHe3Pos", "V0Hyper.fDcaPos2PrimaryVertex"],
-                flatten=True)
+        pytest.importorskip("pandas")
+        t = uproot.open("tests/samples/issue232.root")["fTreeV0"]
+        t.pandas.df(
+            ["V0Hyper.fNsigmaHe3Pos", "V0Hyper.fDcaPos2PrimaryVertex"],
+            flatten=True)
 
     @pytest.mark.skip(reason="This one takes way too long (eospublic?).")
     def test_issue240(self):
-        try:
-            import pyxrootd
-        except ImportError:
-            pytest.skip("unable to import pyxrootd")
-        else:
-            t = uproot.open(
-                "root://eospublic.cern.ch//eos/root-eos/cms_opendata_2012_nanoaod/Run2012B_DoubleMuParked.root"
-            )["Events"]
-            assert (abs(t.array("nMuon")) < 50).all()
+        pytest.importorskip("pyxrootd")
+        t = uproot.open(
+            "root://eospublic.cern.ch//eos/root-eos/cms_opendata_2012_nanoaod/Run2012B_DoubleMuParked.root"
+        )["Events"]
+        assert (abs(t.array("nMuon")) < 50).all()
 
     def test_issue243(self):
         t = uproot.open("tests/samples/issue243.root")["triggerList"]

--- a/tests/test_issues.py
+++ b/tests/test_issues.py
@@ -273,13 +273,12 @@ class Test(object):
             ["V0Hyper.fNsigmaHe3Pos", "V0Hyper.fDcaPos2PrimaryVertex"],
             flatten=True)
 
-    @pytest.mark.skip(reason="This one takes way too long (eospublic?).")
     def test_issue240(self):
         pytest.importorskip("pyxrootd")
         t = uproot.open(
             "root://eospublic.cern.ch//eos/root-eos/cms_opendata_2012_nanoaod/Run2012B_DoubleMuParked.root"
         )["Events"]
-        assert (abs(t.array("nMuon")) < 50).all()
+        assert (abs(t.array("nMuon", entrystop=100000)) < 50).all()
 
     def test_issue243(self):
         t = uproot.open("tests/samples/issue243.root")["triggerList"]

--- a/tests/test_issues.py
+++ b/tests/test_issues.py
@@ -2,10 +2,6 @@
 
 # BSD 3-Clause License; see https://github.com/scikit-hep/uproot/blob/master/LICENSE
 
-import unittest
-
-from collections import namedtuple
-
 import pytest
 import numpy
 
@@ -16,7 +12,7 @@ import uproot_methods.classes.TVector3
 import uproot_methods.classes.TLorentzVector
 
 
-class Test(unittest.TestCase):
+class Test(object):
     def test_issue21(self):
         t = uproot.open("tests/samples/issue21.root")["nllscan"]
 

--- a/tests/test_jagged.py
+++ b/tests/test_jagged.py
@@ -2,15 +2,13 @@
 
 # BSD 3-Clause License; see https://github.com/scikit-hep/uproot/blob/master/LICENSE
 
-import unittest
-
 import uproot
 try:
     import pandas
 except ImportError:
     pandas = None
 
-class Test(unittest.TestCase):
+class Test(object):
     if pandas is not None:
         sample = uproot.open("tests/samples/sample-6.10.05-uncompressed.root")["sample"]
 

--- a/tests/test_jagged.py
+++ b/tests/test_jagged.py
@@ -2,33 +2,35 @@
 
 # BSD 3-Clause License; see https://github.com/scikit-hep/uproot/blob/master/LICENSE
 
+import pytest
+
 import uproot
-try:
-    import pandas
-except ImportError:
-    pandas = None
+
 
 class Test(object):
-    if pandas is not None:
-        sample = uproot.open("tests/samples/sample-6.10.05-uncompressed.root")["sample"]
+    @property
+    def sample(self):
+        pytest.importorskip("pandas")
+        try:
+            self._sample
+        except AttributeError:
+            self._sample = uproot.open("tests/samples/sample-6.10.05-uncompressed.root")["sample"]
+        return self._sample
 
     def test_flatten_False(self):
-        if pandas is not None:
-            df = self.sample.pandas.df(flatten=False)
-            assert len(df.keys()) == 57
-            assert "Af8" in df
-            assert len(df.at[0, "Af8"]) == 0
-            assert len(df.at[1, "Af8"]) == 1
-            assert len(df.at[2, "Af8"]) == 2
+        df = self.sample.pandas.df(flatten=False)
+        assert len(df.keys()) == 57
+        assert "Af8" in df
+        assert len(df.at[0, "Af8"]) == 0
+        assert len(df.at[1, "Af8"]) == 1
+        assert len(df.at[2, "Af8"]) == 2
 
     def test_flatten_None(self):
-        if pandas is not None:
-            df = self.sample.pandas.df(flatten=None)
-            assert len(df.keys()) == 46
-            assert "Af8" not in df
+        df = self.sample.pandas.df(flatten=None)
+        assert len(df.keys()) == 46
+        assert "Af8" not in df
 
     def test_flatten_True(self):
-        if pandas is not None:
-            df = self.sample.pandas.df(flatten=True)
-            assert len(df.keys()) == 57
-            assert "Af8" in df
+        df = self.sample.pandas.df(flatten=True)
+        assert len(df.keys()) == 57
+        assert "Af8" in df

--- a/tests/test_stlvector.py
+++ b/tests/test_stlvector.py
@@ -2,13 +2,9 @@
 
 # BSD 3-Clause License; see https://github.com/scikit-hep/uproot/blob/master/LICENSE
 
-import unittest
-
-from collections import namedtuple
-
 import uproot
 
-class Test(unittest.TestCase):
+class Test(object):
     def runTest(self):
         pass
 

--- a/tests/test_tree.py
+++ b/tests/test_tree.py
@@ -3,12 +3,9 @@
 # BSD 3-Clause License; see https://github.com/scikit-hep/uproot/blob/master/LICENSE
 
 import os
-import unittest
-
 from collections import namedtuple
 
 import numpy
-import pytest
 
 import awkward
 import uproot
@@ -18,7 +15,7 @@ def basest(array):
         array = array.base
     return array
 
-class Test(unittest.TestCase):
+class Test(object):
     ###################################################### double32
 
     def test_double32(self):

--- a/tests/test_tree.py
+++ b/tests/test_tree.py
@@ -6,6 +6,7 @@ import os
 from collections import namedtuple
 
 import numpy
+import pytest
 
 import awkward
 import uproot
@@ -591,9 +592,12 @@ class Test(object):
             for i in range(len(lazy), 0, -1):
                 assert lazy[i - 1 : i + 3].tolist() == strict[i - 1 : i + 3].tolist()
 
-    def test_hist_in_tree(self):
-        path = os.path.join("tests", "samples", "Event.root")
-        if os.path.exists(path):
+    @pytest.mark.parametrize("use_http", [False, True])
+    def test_hist_in_tree(self, use_http):
+        if use_http:
+            path = os.path.join("tests", "samples", "Event.root")
+            if not os.path.exists(path):
+                raise pytest.skip()
             tree = uproot.open(path)["T"]
         else:
             tree = uproot.open("http://scikit-hep.org/uproot/examples/Event.root")["T"]
@@ -607,7 +611,8 @@ class Test(object):
 
         assert tree.array("fH")[20].values.tolist() == check
 
-    def test_branch_auto_interpretation(self):
+    @pytest.mark.parametrize("use_http", [False, True])
+    def test_branch_auto_interpretation(self, use_http):
         # The aim is to reduce this list in a controlled manner
         known_branches_without_interp = [
             b'event',
@@ -619,8 +624,10 @@ class Test(object):
             b'fTriggerBits',
             b'fTriggerBits.TObject'
         ]
-        path = os.path.join("tests", "samples", "Event.root")
-        if os.path.exists(path):
+        if use_http:
+            path = os.path.join("tests", "samples", "Event.root")
+            if not os.path.exists(path):
+                raise pytest.skip()
             tree = uproot.open(path)["T"]
         else:
             tree = uproot.open("http://scikit-hep.org/uproot/examples/Event.root")["T"]
@@ -647,24 +654,20 @@ class Test(object):
         assert a["y"].tolist() == [1, 2, 3, 4, 5]
         assert a["z"].tolist() == [ord("a"), ord("b"), ord("c"), ord("d"), ord("e")]
 
-        try:
-            import pandas
-        except ImportError:
-            pass
-        else:
-            assert tree.pandas.df()["leaflist.x"].tolist() == [1.1, 2.2, 3.3, 4.0, 5.5]
+        pytest.importorskip("pandas")
+        assert tree.pandas.df()["leaflist.x"].tolist() == [1.1, 2.2, 3.3, 4.0, 5.5]
 
-            tree = uproot.open("tests/samples/HZZ-objects.root")["events"]
-            tree.pandas.df("muonp4")
-            tree.pandas.df("muonp4", flatten=False)
-            df = tree.pandas.df("eventweight", entrystart=100, entrystop=200)
-            index = df.index.tolist()
-            assert min(index) == 100
-            assert max(index) == 199
-            df = tree.pandas.df("muonp4", entrystart=100, entrystop=200)
-            index = df.index.get_level_values("entry").tolist()
-            assert min(index) == 100
-            assert max(index) == 199
+        tree = uproot.open("tests/samples/HZZ-objects.root")["events"]
+        tree.pandas.df("muonp4")
+        tree.pandas.df("muonp4", flatten=False)
+        df = tree.pandas.df("eventweight", entrystart=100, entrystop=200)
+        index = df.index.tolist()
+        assert min(index) == 100
+        assert max(index) == 199
+        df = tree.pandas.df("muonp4", entrystart=100, entrystop=200)
+        index = df.index.get_level_values("entry").tolist()
+        assert min(index) == 100
+        assert max(index) == 199
 
     def test_mempartitions(self):
         t = uproot.open("tests/samples/sample-5.23.02-zlib.root")["sample"]

--- a/tests/test_versions.py
+++ b/tests/test_versions.py
@@ -2,8 +2,6 @@
 
 # BSD 3-Clause License; see https://github.com/scikit-hep/uproot/blob/master/LICENSE
 
-import unittest
-
 try:
     import lzma
 except ImportError:
@@ -12,7 +10,7 @@ import lz4
 
 import uproot
 
-class Test(unittest.TestCase):
+class Test(object):
     sample = {
         b"n": [0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4, 0, 1, 2, 3, 4],
 

--- a/uproot/source/http.py
+++ b/uproot/source/http.py
@@ -40,7 +40,7 @@ class HTTPSource(uproot.source.chunked.ChunkedSource):
         while True:
             response = requests.get(
                 self.path,
-                headers={"Range": "bytes={0}-{1}".format(chunkindex * self._chunkbytes, (chunkindex + 1) * self._chunkbytes)},
+                headers={"Range": "bytes={0}-{1}".format(chunkindex * self._chunkbytes, (chunkindex + 1) * self._chunkbytes - 1)},
                 auth=self.auth,
             )
             if response.status_code == 504:   # timeout, try it again

